File: packages/ld-adapter/src/read/get-flag-state.ts
import {
	GetFlagStateInputSchema,
	type FlagEnvironmentState,
	type FlagState,
} from '@ff-promo/contracts';
import type { LaunchDarklyRawClient } from '../client/ld-api-client.js';
import { LdApiError } from '../errors/ld-adapter-error.js';
import { mapLdFlagToFlagState } from './mappers.js';

export type GetFlagStateDeps = {
	client: LaunchDarklyRawClient;
};

function promisifyGetFeatureFlag(
	client: LaunchDarklyRawClient,
	projectKey: string,
	flagKey: string,
): Promise<unknown> {
	return new Promise((resolve, reject) => {
		client.flagsApi.getFeatureFlag(projectKey, flagKey, {}, (error, data) => {
			if (error) {
				reject(error);
				return;
			}
			resolve(data);
		});
	});
}

export function getEnvironmentState(
	flagState: FlagState,
	environmentKey: string,
): FlagEnvironmentState {
	const env = flagState.environments[environmentKey];
	if (!env) {
		throw new LdApiError(
			`Environment not found in flag state: ${environmentKey}`,
			404,
			undefined,
			{
				projectKey: flagState.projectKey,
				flagKey: flagState.flagKey,
				environmentKey,
			},
		);
	}
	return env;
}

export async function getFlagState(
	deps: GetFlagStateDeps,
	input: {
		projectKey: string;
		flagKey: string;
		environmentKey: string;
	},
): Promise<FlagState> {
	const parsed = GetFlagStateInputSchema.parse(input);
	const raw = await promisifyGetFeatureFlag(
		deps.client,
		parsed.projectKey,
		parsed.flagKey,
	);
	const flagState = mapLdFlagToFlagState(
		raw as Parameters<typeof mapLdFlagToFlagState>[0],
		parsed.projectKey,
		parsed.flagKey,
	);
	getEnvironmentState(flagState, parsed.environmentKey);
	return flagState;
}

File: kotlin/modules/ld-adapter/src/main/kotlin/com/ffpromo/ldadapter/read/GetFlagState.kt
package com.ffpromo.ldadapter.read

import com.ffpromo.contracts.FlagEnvironmentState
import com.ffpromo.contracts.FlagState
import com.ffpromo.contracts.GetFlagStateInput
import com.ffpromo.ldadapter.client.LaunchDarklyRawClient
import com.ffpromo.ldadapter.client.LdApiClient
import com.ffpromo.ldadapter.errors.LdApiError

data class GetFlagStateDeps(
    val client: LaunchDarklyRawClient,
)

fun getEnvironmentState(
    flagState: FlagState,
    environmentKey: String,
): FlagEnvironmentState {
    val env = flagState.environments[environmentKey]
        ?: throw LdApiError(
            message = "Environment not found in flag state: $environmentKey",
            status = 404,
            context = mapOf(
                "projectKey" to flagState.projectKey,
                "flagKey" to flagState.flagKey,
                "environmentKey" to environmentKey,
            ),
        )
    return env
}

suspend fun getFlagState(deps: GetFlagStateDeps, input: GetFlagStateInput): FlagState {
    val apiClient = LdApiClient(deps.client)
    val raw = apiClient.getFeatureFlag(input.projectKey, input.flagKey)
    val flagState = mapLdFlagToFlagState(raw, input.projectKey, input.flagKey)
    getEnvironmentState(flagState, input.environmentKey)
    return flagState
}
